### PR TITLE
Engine - Conditional builders refactor

### LINKED ISSUE #13183: Engine - Refactor current builders
We want to give the user the ability to express complex logical expressions to decide whether an event can be processed by an asset, or as a condition to execute a mapping.

The user should be able to express conditions using first-order logic expressions, e.g.:
```
event.category=="registry" AND event.type=="change" AND (
    registry.path =="\SOFTWARE\Policies\Microsoft\Edge\BuiltInDnsClientEnabled"
    OR registry.path=="\SOFTWARE\Google\Chrome\DnsOverHttpsMode"
)
```

This development will deliver a stage builder to implement logic expression checks. This builder will parse the expression and generate an evaluator function to be used in the returned lifter using only one Rxcpp filter observable.

In this issue, Engine builders return a Rxcpp lifter function, while this module requires builders that return the filter function without the Rxcpp wrapper.

  So a little refactor is needed to add a middle builder that returns this function, which will be called inside current builders. These middle builders should be in the Register as well.
  
  Notice that current builders must not change their API nor its tests.

### PR DESCRIPTION
**Pull request by JavierBejMen**
|Related issue|
|---|
|Closes #13183 |

To support conditional expressions a new builder needs to be added, to generate the Lifter rxcpp function from a json check definition. This builder needs all conditional builders to return the filter function instead the Lifter, so this builders has been refactored to do so, being the opConditionBuilder the one that creates the Lifter for all conditional builders. Middle helper builders has been added to the Registry too.

Changes:
- Added stageExpressionCheckBuilder with test.
- Integrated opConditionValueBuilder and opConditionReferenceBuilders into opConditionBuilder.
- Refactored all conditional builders to return a function `bool(Event)`, delegating Lifter creation to subsequent builders.
- Updated tests.

## Tests (Local)
image

### CODE REVIEW

**Review by juliancnn — APPROVED** [review]
LGTM!

File: src/engine/source/builder/builders/opBuilderCondition.cpp [hunk 1]
```diff
@@ -19,53 +20,145 @@
...
+                                 "definition must be a string.");
+    }
 
-types::Lifter opBuilderCondition(const types::DocumentValue & def, types::TracerFn tr)
+    std::string field =
+        json::formatJsonPath(def.MemberBegin()->name.GetString());
+    // TODO: build document with value only
+    types::Document value {def};
+    std::string successTrace = fmt::format("{} Condition Success", value.str());
+    std::string failureTrace = fmt::format("{} Condition Failure", value.str());
```
Comment (nmkoremblum) [inline src/engine/source/builder/builders/opBuilderCondition.cpp:85]: Please, make them "const"

File: src/engine/source/builder/builders/opBuilderCondition.cpp [hunk 2]
```diff
@@ -19,53 +20,145 @@
...
+    }
+    reference = json::formatJsonPath(reference);
+    std::string successTrace =
+        fmt::format("{{{}: {}}} Condition Success",
+                    def.MemberBegin()->name.GetString(),
+                    def.MemberBegin()->value.GetString());
+    std::string failureTrace =
+        fmt::format("{{{}: {}}} Condition Failure",
+                    def.MemberBegin()->name.GetString(),
+                    def.MemberBegin()->value.GetString());
```
Comment (nmkoremblum) [inline src/engine/source/builder/builders/opBuilderCondition.cpp:54]: Please, make them "const"

File: src/engine/source/builder/builders/opBuilderCondition.cpp [hunk 3]
```diff
@@ -19,53 +20,145 @@
...
+    if (!def.MemberBegin()->value.IsString())
+    {
+        throw std::runtime_error("Error building condition reference, value of "
+                                 "definition must be a string.");
+    }
+
+    // Estract and prepare field and reference
+    std::string field {
+        json::formatJsonPath(def.MemberBegin()->name.GetString())};
+    std::string reference {def.MemberBegin()->value.GetString()};
```
Comment (nmkoremblum) [inline src/engine/source/builder/builders/opBuilderCondition.cpp:41]: Please, make them "const"

File: src/engine/source/builder/builders/stageBuilderExpressionCheck.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+            field = term.substr(pos1 + 1, pos2);
+            value = term.substr(0, pos1) + term.substr(pos2, term.size());
+        }
+
+        types::Document doc;
+        doc.m_doc.SetObject();
+        doc.m_doc.GetObject().AddMember({field.c_str(), doc.getAllocator()},
+                                        {value.c_str(), doc.getAllocator()},
+                                        doc.getAllocator());
```
Comment (nmkoremblum) [inline src/engine/source/builder/builders/stageBuilderExpressionCheck.cpp:64]: It would be nice to have a comment here to explain why is this necessary

File: src/engine/source/builder/builders/stageBuilderExpressionCheck.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+        auto termFunction = std::get<types::MiddleBuilderCondition>(
+            Registry::getBuilder("middle.condition"))(doc.getObject(), tr);
+        return termFunction;
```
Comment (nmkoremblum) [inline src/engine/source/builder/builders/stageBuilderExpressionCheck.cpp:68]: Can't we here just return the "middle.condition" builder method?

Comment (nmkoremblum) [reply src/engine/source/builder/builders/stageBuilderExpressionCheck.cpp:68]: (middleBuilderCondition)

File: src/engine/test/source/builder/builders/assetBuilderOutput_test.cpp [hunk 6]
```diff
@@ -107,28 +104,29 @@ TEST(AssetBuilderOutput, BuildsOperates)
...
                 {"field":"value1"}
             )"));
-            s.on_next(std::make_shared<json::Document>(R"(
+                         s.on_next(std::make_shared<json::Document>(R"(
                 {"field":"value"}
             )"));
-            s.on_next(std::make_shared<json::Document>(R"(
+                         s.on_next(std::make_shared<json::Document>(R"(
                 {"field":"value"}
             )"));
```
Comment (nmkoremblum) [inline src/engine/test/source/builder/builders/assetBuilderOutput_test.cpp:126]: Is this correctly idented?

File: src/engine/test/source/builder/builders/assetBuilderOutput_test.cpp [hunk 7]
```diff
@@ -143,8 +141,8 @@ TEST(AssetBuilderOutput, BuildsOperates)
 {"field":"value"}
 )";
 
-    string filepath1{"/tmp/stageOutputsTest1.txt"};
-    string filepath2{"/tmp/stageOutputsTest2.txt"};
+    string filepath1 {"/tmp/stageOutputsTest1.txt"};
+    string filepath2 {"/tmp/stageOutputsTest2.txt"};
```
Comment (nmkoremblum) [inline src/engine/test/source/builder/builders/assetBuilderOutput_test.cpp:145]: Isn't it better to have a "define" for this?

**Review by nmkoremblum — CHANGES_REQUESTED** [review]

File: src/engine/source/builder/builders/opBuilderCondition.cpp (hunk 2 above)
Comment (JavierBejMen) [reply src/engine/source/builder/builders/opBuilderCondition.cpp:54]: done

File: src/engine/source/builder/builders/opBuilderCondition.cpp (hunk 3 above)
Comment (JavierBejMen) [reply src/engine/source/builder/builders/opBuilderCondition.cpp:41]: only made const field, reference is edited just in the following line

File: src/engine/test/source/builder/builders/assetBuilderOutput_test.cpp (hunk 6 above)
Comment (JavierBejMen) [reply src/engine/test/source/builder/builders/assetBuilderOutput_test.cpp:126]: Done with latest format, we may need to fix it

File: src/engine/test/source/builder/builders/assetBuilderOutput_test.cpp (hunk 7 above)
Comment (JavierBejMen) [reply src/engine/test/source/builder/builders/assetBuilderOutput_test.cpp:145]: May be, not on scope on this PR

File: src/engine/source/builder/builders/stageBuilderExpressionCheck.cpp (hunk 5 above)
Comment (JavierBejMen) [reply src/engine/source/builder/builders/stageBuilderExpressionCheck.cpp:68]: Added a todo, to be debated on upcoming Engine meeting.

**Review by nmkoremblum — APPROVED** [review]
LGTM
